### PR TITLE
Add `no-index-component-invocation` rule

### PR DESCRIPTION
**Pull request by lifeart**
https://github.com/ember-template-lint/ember-template-lint/issues/894

should it be in octane config?

// cc @rwjblue, @NullVoxPopuli

### CODE REVIEW

File: lib/rules/lint-no-index-component-invocation.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+module.exports = class LintNoIndexComponentInvocation extends Rule {
+  visitor() {
+    const sourceForNode = this.sourceForNode.bind(this);
+    const log = this.log.bind(this);
+    function lintIndexUsage(node) {
+      if (node.type === 'ElementNode') {
+        if (node.tag.endsWith('::Index')) {
+          return log({
+            message: `Replace "<${node.tag} ..." to <${node.tag.replace('::Index', '')} ..."`,
```
[the remark this answers is not in the document]
Comment (lifeart) [reply lib/rules/lint-no-index-component-invocation.js]: Good catch!

File: test/unit/rules/lint-no-index-component-invocation-test.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        message: 'Replace "{{foo/index ..." to "{{foo ..."',
+        source: 'foo/index',
+        line: 1,
+        column: 2,
+      },
+    },
+    {
+      template: '{{foo/bar (component "foo/index")}}',
+      result: {
+        message: 'Replace "(component "foo/index" ..." to "(component "foo" ..."',
```
[the remark this answers is not in the document]
Comment (lifeart) [reply test/unit/rules/lint-no-index-component-invocation-test.js]: Done

**Comment by lifeart** [comment]
//cc @rwjblue

File: docs/rule/no-index-component-invocation.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+## no-index-component-invocation
+
+We propose to allow `index`
+files inside a directory to have the equivalent semantics. Components and Component Templates can be structured as `app/components/foo-bar/index.js` and
+`app/components/foo-bar/index.hbs`. This allows additional files related to the
+component (such as a `README.md` file) to be co-located on the filesystem.
```
Comment (Turbo87) [inline docs/rule/no-index-component-invocation.md]: While this is possible, it is not the default for Ember projects so IMHO we shouldn't be suggesting it here. We can mention that this works, but "propose" might be the wrong word here.

File: lib/rules/lint-no-index-component-invocation.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+const Rule = require('./base');
+
+module.exports = class LintNoIndexComponentInvocation extends Rule {
+  visitor() {
+    const sourceForNode = this.sourceForNode.bind(this);
+    const log = this.log.bind(this);
+    function lintIndexUsage(node) {
```
Comment (Turbo87) [inline lib/rules/lint-no-index-component-invocation.js]: if you make this a method on the `LintNoIndexComponentInvocation` class you don't have to explicitly bind the methods above 😉

Comment (lifeart) [reply lib/rules/lint-no-index-component-invocation.js]: using function make easier to test logic in astexplorer

File: docs/rule/no-index-component-invocation.md (hunk 3 above)
Comment (lifeart) [reply docs/rule/no-index-component-invocation.md]: Tobias, could you provide more suitable sentence?

**Comment by lifeart** [comment]
image

**Comment by MrChocolatine** [comment]
Does this rule only apply to Octane or what ? I am running a project with Ember 3.12 and if I rename a component `<Foo::Index />` to `<Foo />` it does not seem to be rendered anymore.

Actually in Ember Inspector I can see `<Foo />` its render is blank. If I keep the `::Index` the component renders correctly. What could be the issue ?

**Comment by lifeart** [comment]
@MrChocolatine is component template located into `templates/components/foo/index.hbs` ?

**Comment by MrChocolatine** [comment]
> @MrChocolatine is component template located into `templates/components/foo/index.hbs` ?

It is yes.

The actual files of my component are:
```
app/components/foo/bar/index.js
app/templates/components/foo/bar/index.hbs
```

When I invoke this component:
```hbs
-- works --
<Foo::Bar::Index @model={{this.model.whatever}} />

-- does not work / render nothing --
<Foo::Bar @model={{this.model.whatever}} />
```

**Comment by lifeart** [comment]
@MrChocolatine could you try update ember-cli-babel version?

**Comment by MrChocolatine** [comment]
I just updated it to its last version `ember-cli-babel@7.20.5` and I still have the same issue.

I even tried to simplify the process but it didn't help:
- A simple text in the component's HBS file
- Empty content of the component's JS class.
- Change invocation of the component `<Foo::Bar />` (without argument this time)

If I delete the component's JS file I get an error `Compile Error: Cannot find component foo/bar`.

**Comment by lifeart** [comment]
@MrChocolatine yeah, it's related to template compilation logic updates (that's why it's working for octane)

cc  @rwjblue

**Comment by MrChocolatine** [comment]
That is what I expected.
When I started to make components with `::Index` I tried to get rid of it without success. It in the end, if I remember correctly, I had to upgrade my Ember version to make it works without `::Index`. (and I will, but not immediately)

**Comment by MrChocolatine** [comment]
We recently did a massive upgrade of dependencies to their latest version, except for `ember-cli`, `ember-source` and `ember-data` but nothing changed, `<Foo::Bar />` still renders nothing.

Still convinced this rule should only apply to Octane projects.

**Comment by rwjblue** [comment]
You definitely do not need to be using Octane to invoke nested components like this (or to use colocation in general).

**Comment by MrChocolatine** [comment]
> You definitely do not need to be using Octane to invoke nested components like this (or to use colocation in general).

We are speaking of `<Foo::Bar::Index />` and `<Foo::Bar />` right ? (I just edited my previous message)
Then I don't see what could the issue since I am using Ember 3.12 and every other dependencies are up-to-date 🤔 .

**Comment by lifeart** [comment]
As I understand issue happens on such layout:
```
app/components/foo/index.js
app/templates/components/foo/index.hbs
```

**Pull request merged**